lib/src/strongly_connected_components.dart

```diff
@@ -67,7 +75,7 @@ List<List<T>> stronglyConnectedComponents<T>(
         next = lastVisited.removeLast();
         onStack.remove(next);
         component.add(next);
-      } while (!equals(next, node));
+      } while (!equals!(next, node));
```

Comment:
I am a bit confused why we need the `equals!` here? It should be promoted to non_nullable by this line above? `equals ??= _defaultEquals;`

Reply:
This is one of those known, gnarly issues with promotion w/ inline functions, I'm pretty sure.

CC @leafpetersen @munificent

Reply:
one might could create another local, final var to store this so we don't need to do the thing.

Reply:
cc @stereotype441 is this because this is a local function and we don't realize it isn't invoked until after that is promoted or would you expect this to work?

Reply:
Oh didn't see your comment yet lol, guess we are pinging all the people lol :)

Reply:
@kevmoo is right that it's one of those known gnarly promotion issues.  Flow analysis sees that a write to `equals` exists *somewhere* in the source code, and therefore it thinks it's not safe to carry the promotion into the local function declaration, because maybe sometime after the local function is declared, but before it is called, a `null` value *might* get written into `equals`.

It's not smart enough to see that in this case (a) `equals` never gets written again, and (b) even if it did, the `??=` can never cause it to be `null`.

It all kind of comes back to our decision to make flow analysis do all of its reasoning in two passes over the source code rather than iterating to a fixed point.  I still think that was the right call overall (because it keeps compilation fast and reduces the risk of weird action at a distance), but it does mean that we miss cases like this one.